docs/agda-spec/src/Spec/OperationalCertificate.lagda

```diff
@@ -0,0 +1,141 @@
+  where
+
+open import Spec.BaseTypes crypto using (OCertCounters)
+open import Ledger.Prelude
+
+\end{code}
+
+\begin{figure*}[h]
+\emph{Operational Certificate environments}
+\begin{code}
+OCertEnv = ℙ KeyHashˢ
+\end{code}
+\emph{Operational Certificate states}
+\begin{code}
+OCertState = OCertCounters
+\end{code}
+\emph{Operational Certificate transitions}
+\begin{code}[hide]
+data
+\end{code}
+\begin{code}
+  _⊢_⇀⦇_,OCERT⦈_ : OCertEnv → OCertState → BHeader → OCertState → Type
+\end{code}
+\emph{Operational Certificate helper function}
+\begin{code}
+currentIssueNo : OCertEnv → OCertState → KeyHashˢ → Maybe ℕ
+currentIssueNo stpools cs hk =
+  if hk ∈ dom (cs ˢ) then
+    just (lookupᵐ cs hk)
+  else
+  if hk ∈ stpools then
+    just 0
+  else
+    nothing
+\end{code}
+\caption{Operational Certificate transition-system types and functions}
+\label{fig:ts-types:ocert}
+\end{figure*}
+
+The transition rule $\mathsf{OCERT}$ is shown in Figure~\ref{fig:ts-rules:ocert}. From the block
+header body \aarg{bhb} we first extract the following:
+
+\begin{itemize}
+  \item The operational certificate \afun{oc}, consisting of the hot key \aarg{vkₕ},
+    the certificate issue number \aarg{n}, the KES period start \aarg{c₀} and the cold key
+  signature \aarg{τ}.
+\item The cold key \afun{issuerVk}.
+\item The slot \afun{slot} for the block.
+\item The number \aarg{t} of KES periods that have elapsed since the start period on the certificate.
+\end{itemize}
+
+Using this we verify the preconditions of the operational certificate state
+transition which are the following:
+
+\begin{itemize}
+\item The KES period \aarg{kp} of the slot in the block header body must be greater than or equal to
+  the start value \aarg{c₀} listed in the operational certificate,
+  and less than \afld{MaxKESEvo}-many KES periods after \aarg{c₀}.
+  The value of \afld{MaxKESEvo} is the agreed-upon lifetime of an operational certificate,
+  see \cite{delegation_design}.
+\item \aarg{hk} exists as key in the mapping of certificate issues numbers to a KES
+  period \aarg{m} and that period is less than or equal to \aarg{n}. Also, \aarg{n} must
+  be less than or equal to the successor of \aarg{m}.
+\item The signature \aarg{τ} can be verified with the cold verification key \afun{issuerVk}.
+\item The KES signature \aarg{σ} can be verified with the hot verification key \aarg{vkₕ}.
+\end{itemize}
+
+After this, the transition system updates the operational certificate state by
+updating the mapping of operational certificates where it overwrites the entry
+of the key \aarg{hk} with the KES period \aarg{n}.
+
+\begin{figure*}[h]
+\begin{code}[hide]
+private variable
+  stpools : OCertEnv
+  cs      : OCertState
+  bh      : BHeader
+
+data _⊢_⇀⦇_,OCERT⦈_ where
+\end{code}
+\begin{code}
+  Update-OCert :
+    let (bhb , σ) = bh; open BHBody bhb
+        ⟦ vkₕ , n , c₀ , τ ⟧ᵒᶜ = oc
+        hk = hash issuerVk
+        kp = kesPeriod slot
+        t = kp -ᵏ c₀

```

Comment:
We have `HasAdd` and `HasSubtract` type classes, so you could just use regular `+` and `-` operations for these.

Reply:
Actually, I didn't use such type classes because the signatures do not match: For instance, `-ᵏ` has type `KESPeriod → KESPeriod → ℕ` instead of `KESPeriod → KESPeriod → KESPeriod` as required by `HasSubtract`.